lib/network/rpc/rpc.js

```diff
@@ -0,0 +1,475 @@
+        this.streamAckTimeoutMs = streamAckTimeoutMs;
+
+        this.requestInfoProducers = [];
+        this.requestInfoProducers.push(
+            dbClient => ({ reqUids: dbClient.withReqUids }));
+    }
+
+    /**
+     * @brief internal RPC implementation w/o timeout
+     *
+     * @param {String} remoteCall - name of the remote function to call
+     * @param {Array} args - list of arguments to the remote function
+     * @param {function} cb - callback called when done
+     * @return {undefined}
+     */
+    _call(remoteCall, args, cb) {
+        const wrapCb = (err, data) => {
+            cb(reconstructError(err),
+               this.socketStreams.decodeStreams(data));
+        };
+        this.logger.debug('remote call', { remoteCall, args });
+        this.socket.emit('call', remoteCall,
+                         this.socketStreams.encodeStreams(args), wrapCb);
+        return undefined;
+    }
+
+    /**
+     * @brief call a remote function named @a remoteCall, with
+     * arguments @a args and callback @a cb
+     *
+     * @a cb is called when the remote function returns an ack, or
+     * when the timeout set by @a timeoutMs expires, whichever comes
+     * first. When an ack is received, the callback gets the arguments
+     * sent by the remote function in the ack response. In the case of
+     * timeout, it's passed a single Error argument with the code:
+     * 'ETIMEDOUT' property, and a self-described string in the 'info'
+     * property.
+     *
+     * @param {String} remoteCall - name of the remote function to call
+     * @param {Array} args - list of arguments to the remote function
+     * @param {function} cb - callback called when done or timeout
+     * @param {Number} timeoutMs - timeout in milliseconds
+     * @return {undefined}
+     */
+    callTimeout(remoteCall, args, cb, timeoutMs = DEFAULT_CALL_TIMEOUT_MS) {
+        if (typeof cb !== 'function') {
+            throw new Error(`argument cb=${cb} is not a callback`);
+        }
+        async.timeout(this._call.bind(this), timeoutMs,
+                      `operation ${remoteCall} timed out`)(remoteCall,
+                                                           args, cb);
+        return undefined;
+    }
+
+    getCallTimeout() {
+        return this.callTimeoutMs;
+    }
+    setCallTimeout(newTimeoutMs) {
+        this.callTimeoutMs = newTimeoutMs;
+    }
+
+    /**
+     * connect to the remote RPC server
+     *
+     * @param {function} cb - callback when connection is complete or
+     *   if there is an error
+     * @return {undefined}
+     */
+    connect(cb) {
+        this.socket = ioClient(this.url);
+        this.socketStreams = sioStream.createSocket(
+            this.socket,
+            this.logger,
+            this.streamMaxPendingAck,
+            this.streamAckTimeoutMs);
+        this.socket.on('error', err => {
+            this.logger.warn('connectivity error to storage daemon',
```

Comment:
Now I think I get what you mean, because it's in the common RPC layer, it it not necessarily talking to a storage daemon. That is right, even though for now this is what we use it for.

I could change the message to "connectivity error to the RPC server", adding the URL to the error params so that people know where it goes to.